src/ucs/arch/generic/cpu.c

```diff
@@ -0,0 +1,109 @@
+
+#define UCS_CPU_CACHE_FILE_FMT     "/sys/devices/system/cpu/cpu%d/cache/index%d/%s"
+#define UCS_CPU_CACHE_LEVEL        "level"
+#define UCS_CPU_CACHE_TYPE         "type"
+#define UCS_CPU_CACHE_SIZE         "size"
+
+static ucs_init_once_t ucs_cache_read_once = UCS_INIT_ONCE_INIITIALIZER;
+
+/* cache datatypes */
+struct { /* cache info obtained from [linux] kernel */
+    int             level;
+    const char     *type;
+} const ucs_cpu_cache_kernel[] = {
+    [UCS_CPU_CACHE_L1d] = {.level = 1, .type = "Data"},
+    [UCS_CPU_CACHE_L1i] = {.level = 1, .type = "Instruction"},
+    [UCS_CPU_CACHE_L2]  = {.level = 2, .type = "Unified"},
+    [UCS_CPU_CACHE_L3]  = {.level = 3, .type = "Unified"}
+};
+
+ucs_cache_size_t ucs_cpu_cache_size[] = { /* cached values of CPU cache */
+    [UCS_CPU_CACHE_L1d] = {.cache_size = 0, .status = UCS_ERR_UNSUPPORTED},
+    [UCS_CPU_CACHE_L1i] = {.cache_size = 0, .status = UCS_ERR_UNSUPPORTED},
+    [UCS_CPU_CACHE_L2]  = {.cache_size = 0, .status = UCS_ERR_UNSUPPORTED},
+    [UCS_CPU_CACHE_L3]  = {.cache_size = 0, .status = UCS_ERR_UNSUPPORTED}
+};
+
+
+static void ucs_cpu_read_cache_size(ucs_cache_size_t *cache, size_t length)
+{
+    char type_str[32];  /* Data/Instruction/Unified */
+    char size_str[32];  /* memunits */
+    int i;
+    int cpu;
+    long level;
+    ssize_t file_size;
+    size_t c;
+    ucs_status_t status;
+
+    /* try first CPU-specific algorithm */
+    ucs_arch_get_cache_size(cache, length);
+
+    ucs_assert(length <= ucs_static_array_size(ucs_cpu_cache_kernel));
+
+    cpu = ucs_get_first_cpu();
+
+    for (i = 0;; i++) {
+        file_size = ucs_read_file_str(type_str, sizeof(type_str), 1, UCS_CPU_CACHE_FILE_FMT,
+                                      cpu, i, UCS_CPU_CACHE_TYPE);
+        if (file_size < 0) {
+            return; /* no more files */
+        }
+
+        status = ucs_read_file_number(&level, 1, UCS_CPU_CACHE_FILE_FMT,
+                                      cpu, i, UCS_CPU_CACHE_LEVEL);
+        if (status != UCS_OK) {
+            return; /* no more files */
+        }
+
+        /* ok, we found valid directory, let's try to read cache size */
+        file_size = ucs_read_file_str(size_str, sizeof(size_str), 1, UCS_CPU_CACHE_FILE_FMT,
+                                      cpu, i, UCS_CPU_CACHE_SIZE);
+        if (file_size < 0) {
+            return; /* no more files */
+        }
+
+        /* now lookup for cache size entry */
+        for (c = 0; c < length; c++) {
+            if ((ucs_cpu_cache_kernel[c].level == level) &&
+                !strcasecmp(ucs_cpu_cache_kernel[c].type, ucs_strtrim(type_str))) {
+                if (cache[c].status != UCS_OK) {
+                    cache[c].status = ucs_str_to_memunits(ucs_strtrim(size_str), &cache[c].cache_size);
+                } else {
+                    break;
+                }
+            }
+        }
+    }
+}
+
+ucs_status_t ucs_cpu_get_cache_size(ucs_cpu_cache_type_t type, size_t *value)
```

Comment:
ucs_arch_generic_get_cache_size

Reply:
renamed to ucs_sysfs_get_cache_size.
arch_get_cache_size is called from ucs_cpu_get_cache_size